source/common/http/http1/codec_impl.h

```diff
@@ -197,6 +188,163 @@ class RequestEncoderImpl : public StreamEncoderImpl, public RequestEncoder {
+   *
+   * This method is preferred instead of directly using the `reserveBuffer` and `writeTobuffer`
+   * methods.
+   *
+   * @param args variable-length argument pack to write.
+   * @return the size of the data written.
+   */
+  template <class... Args> size_t reserveAndWrite(Args&&... args) {
+    size_t size_to_write = 0;
+    if constexpr (sizeof...(args) > 0) {
+      size_to_write = (bytesSize(args) + ...);
+      reserveBuffer(size_to_write);
+      (writeToBuffer(args), ...);
+    }
+    return size_to_write;
+  }
+
+  /**
+   * Write data with specified size to the buffer.
+   * @param data the address of the data to be written.
+   * @param size the size of the data to be written.
+   */
+  void writeToBuffer(const char* data, uint64_t size) {
+    ASSERT(remaining() >= size);
+    memcpy(current_pos_, data, size); // NOLINT(safe-memcpy)
+    current_pos_ += size;
+  }
+
+  /**
+   * Write data without specified size to the buffer.
+   * @param data the address of the data to be written. Copy data to buffer until the first '\0'
+   * character is encountered.
+   */
+  void writeToBuffer(const char* data) {
+    const size_t size = bytesSize(data);
+    writeToBuffer(data, size);
+  }
+
+  /**
+   * Write string view the buffer.
+   * @param data the string view to be written.
+   */
+  void writeToBuffer(absl::string_view data) { writeToBuffer(data.data(), data.size()); }
+
+  /**
+   * Write single char to buffer.
+   * @param c the single char to be written.
+   */
+  void writeToBuffer(char c) {
+    ASSERT(remaining() >= 1);
+    *current_pos_ = c;
+    current_pos_++;
+  }
+
+  /**
+   * Reserve enough local buffer. After that, the data can be copied directly to the buffer without
+   * additional checking. Before calling `writeToBuffer`, this method should be called to ensure
+   * that the local buffer has enough memory space. The common scenario is to call this method once
+   * to reserve a big enough buffer, and then call writeToBuffer multiple times.
+   *
+   * @param size local buffer size to reserve.
+   */
+  void reserveBuffer(uint64_t size) {
+    ASSERT(size > 0);
+
+    if (reservation_ == nullptr) {
+      initializeReservation(size);
+      return;
+    }
+
+    if (remaining() >= size) {
+      return;
+    }
+
+    commitToBuffer();
+    initializeReservation(size);
+  }
+
+  /**
+   * Write all the contents in the local buffer to the backend buffer.
+   */
+  void commitToBuffer() {
+    if (reservation_ == nullptr) {
+      return;
+    }
+
+    size_t filled = hasFilled();
+
+    if (filled > 0) {
+      reservation_->commit(filled);
+    }
+
+    reservation_ = nullptr;
+    raw_slice_ = {nullptr, 0};
+    current_pos_ = nullptr;
+  }
+
+  // Method only used for test.
+  uint64_t remainingForTest() { return remaining(); }
+
+private:
+  template <class T> size_t bytesSize(T&& arg) {
+    if constexpr (std::is_same_v<char, std::remove_reference_t<T>>) {
```

Comment:
I suspect that all this type gymnastics is not needed if we let string_view do the heavy lifting for us. string_view is constant-time to populate from a std::string or a literal, and will take care of the strlen call if you give it a null-terminated `char*.`

Reply:
This is true if the single char is not considered.

Reply:
>  I just want to centralize this in Buffer's API and not in the codec.

@jmarantz  Right. If we decide to keep this helper in the end, I will move the related code. 

But now I am thinking about can we create new template method and virtual method in the Buffer::Instance make this more clean and simple. In this way, the relevant logic can be implemented directly in `Buffer::Instance`. 

But I haven't figured out how to deal with watermark. Should we check it in the `inlineReserve`. cc @antoniovicente 

```
class Instance {
public:
  // ......

  template <class... Args> size_t addFragments(Args&&... args) {
    size_t size_to_write = 0;
    if constexpr (sizeof...(args) > 0) {
      size_to_write = (bytesSize(args) + ...);
      auto* to_write = inlineReserve(size_to_write);
      (write(to_write, args), ...);
    }
    return size_to_write;
  }

private:

  /**
   * Reserve Reserve a section of contiguous memory, and this section of memory is deemed to have
   * been used. This method directly increases the size of the buffer. Finally, the first address of
   * this memory is returned.
   */
  virtual char* inlineReserve(size_t size) PURE;
};

```